load_test/xmpp_amqp.erl

```diff
@@ -0,0 +1,621 @@
+-spec bind_queue_to_presence_exchange(ChannelPid, Queue, NeighborIDs) ->
+    ok | error when
+      ChannelPid  :: pid(),
+      Queue       :: binary(),
+      NeighborIDs :: [amoc_scenario:user_id()].
+bind_queue_to_presence_exchange(Channel, Queue, NeighborIDs) ->
+    Exchange = amoc_config:get(rabbit_presence_exchange, <<"presence">>),
+    RoutingKeys =
+      [routing_keys_for_user(presence, ID, all_topics) || ID <- NeighborIDs],
+    Results =
+      [bind_queue_to_exchange(Channel, Queue, Exchange, RoutingKey)
+       || RoutingKey <- RoutingKeys],
+    return_ok_if_all_ok_else_error(Results).
+
+% @doc Generic function for binding queue to exchange with given routing key.
+-spec bind_queue_to_exchange(ChannelPid  :: pid(),
+                             Queue       :: binary(),
+                             Exchange    :: binary(),
+                             RoutingKey  :: binary()) -> ok | error.
+bind_queue_to_exchange(Channel, Queue, Exchange, RoutingKey) ->
+    Binding = #'queue.bind'{queue = Queue,
+                            exchange = Exchange,
+                            routing_key = RoutingKey},
+    case amqp_channel:call(Channel, Binding) of
+        #'queue.bind_ok'{} ->
+            ok;
+        Err ->
+            lager:error("Could not bind queue ~p to exchange ~p error: ~p", [Queue, Exchange, Err]),
+            error
+    end.
+
+%================================================
+% XMPP helpers
+%================================================
+
+-spec make_message(binjid()) -> exml:element().
+make_message(ToId) ->
+    Body = <<"hello sir, you are a gentelman and a scholar.">>,
+    Id = escalus_stanza:id(),
+    escalus_stanza:set_id(escalus_stanza:chat_to(ToId, Body), Id).
+
+-spec make_full_jid(amoc_scenario:user_id()) -> binjid().
+make_full_jid(Id) ->
+    BinInt = integer_to_binary(Id),
+    ProfileId = <<"user_", BinInt/binary>>,
+    Host = ?HOST,
+    << ProfileId/binary, "@", Host/binary, "/", ?XMPP_RESOURCE/binary >>.
+
+-spec make_bare_jid(amoc_scenario:user_id()) -> binjid().
+make_bare_jid(Id) ->
+    BinInt = integer_to_binary(Id),
+    ProfileId = <<"user_", BinInt/binary>>,
+    Host = ?HOST,
+    << ProfileId/binary, "@", Host/binary >>.
+
+%================================================
+% RabbitMQ helpers
+%================================================
+
+% @doc Functions starts ?RABBIT_CONNECTIONS connections
+ensure_amqp_connection_pool() ->
+    Host = amoc_config:get(rabbit_host),
+    Port = amoc_config:get(rabbit_port),
+    Username = amoc_config:get(rabbit_username),
+    Password = amoc_config:get(rabbit_password),
+    Params = #amqp_params_network{host = Host,
+                                  port = Port,
+                                  username = Username,
+                                  password = Password},
+    [connect_amqp(Params) || _ <- lists:seq(1, ?RABBIT_CONNECTIONS)].
+
+% @doc This helper function utilizes internal amqp_client supervisor as
```

Comment:
This function really took me by surprise. I think is't a little bit hacky in a way 😃 Maybe we can get rid of it by storing AMQP connections in an application configuration (we can achieve this using `application:set_env/3`) or in a dedicated ETS table? I believe it's worth considering because it'll definitely make the code easier to understand.